355 mergecubes uses default geotrellis partitioner in certain cases which causes oom errors

JeroenVerstraelen opened the pull request:
The default SpacePartitioner uses 16x16 tiles per partition which requires too much memory if there are many bands, dates or if the tiles are large. This commit introduces a partitioner that attempts to control the partition size (max records) based on a maximum memory footprint.

The new partitioner in this special case can have 1,2,4,8,16,32,64,128,256,512, or 1024 records in a single partition based on the dimensions of a single MultiBandTile: tile size, number of bands, and celltype.

JeroenVerstraelen committed:
add ConfigurableSpatialPartitionerReduceZ for more granular partition sizes #355

JeroenVerstraelen committed:
replace default partitioner with getPartitionerIndexForMaxPartitionSize in merge_cubes #355

The default SpacePartitioner[K](kb) uses 16x16 tiles which uses too much memory if there are many bands, dates or if the tiles are large. This commit introduces a partitioner that attempts to control the partition size based on a maximum memory footprint.

jdries commented on openeo-geotrellis/src/main/scala/org/openeo/geotrellis/OpenEOProcesses.scala:
```diff
@@ -708,14 +719,13 @@ class OpenEOProcesses extends Serializable {
     } else {
+      // At least one partitioner is undefined.
       logger.info(s"Merging cubes with partitioners: ${leftCube.partitioner} - ${rightCube.partitioner} - many band case detected: $manyBands")
-      if(manyBands) {
```
why the immediate removal of manyBands?
I would prefer a more incremental change that takes things one step at a time, allowing to release this one faster without too much worries.

JeroenVerstraelen replied:
ManyBands returned either ByTileSpatialPartitioner or SpaceTimeByMonthPartitioner which are both special cases of the configurable partitioners so I thought it was safe. But of course the new partitioner has to be more thoroughly tested to ensure it doesn't actually exceed its maximum allowed memory size so I will revert the manyBands case.

JeroenVerstraelen replied:
Pushed a fix.

jdries commented on openeo-geotrellis/src/main/scala/org/openeo/geotrellis/OpenEOProcesses.scala:
```diff
@@ -708,14 +719,13 @@ class OpenEOProcesses extends Serializable {
-
+      val nrBands = leftCount.getOrElse(1) + rightCount.getOrElse(1)
+      val outputCellType = maybeCellType(leftCube).getOrElse(DoubleCellType).union(maybeCellType(rightCube).getOrElse(DoubleCellType))
+      val tileSize = maybeTileSize(leftCube).getOrElse(128)
```
this is wrong: geotrellis tilesize is actually returning width*height, so something needs to change

jdries replied:
checking out usage of tilesize in getPartitionerIndexForMaxPartitionSize, the default should perhaps be 128*128
some scaladoc might also help to make sure that interpretation is clear

JeroenVerstraelen replied:
Pushed a fix.

JeroenVerstraelen committed:
fix tilesize default in merge cubes #355

JeroenVerstraelen committed:
add documentation to getPartitionerIndexForMaxPartitionSize #355

JeroenVerstraelen committed:
revert manyBands case in outerJoin #355

JeroenVerstraelen committed:
use worst case as default for number of bands #355

With this default, for a maximum partition size of 500 Mb the following would be true:
* 20 bands - 500Mb partitions
* 40 bands - 1Gb partitions
* 10 bands - 250Mb partitions

This should be a good balance, since 40 bands is quite extreme.
